File: apps/desktop-tauri/product/harbor-python/src/harbor_dsh_evolution/candidate_runtime.py
"""A Candidate-owned, content-addressed ACP launch contract.

The descriptor and npm lockfile are Candidate source, not Host defaults. Reading
old snapshots remains supported; executing one requires an explicit binding.
"""
from __future__ import annotations

import base64
import hashlib
import json
import re
from pathlib import Path, PurePosixPath
from typing import Any
from urllib.parse import urlsplit

DESCRIPTOR_NAME = "candidate-runtime.json"
LOCKFILE_NAME = "package-lock.json"
_FIELDS = {"schema_version", "transport", "entrypoint", "config_path", "agent_entry_id", "node_version"}
_SEMVER = re.compile(r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*)(?:\.(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*))*)?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$", re.ASCII)
_NODE_VERSION = re.compile(r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$", re.ASCII)
_ENTRY_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_DEPENDENCIES = ("dependencies", "optionalDependencies", "devDependencies", "peerDependencies")


def _invalid(message: str) -> None:
    raise ValueError(f"CANDIDATE_RUNTIME_INVALID: {message}")


def _file(root: Path, relative: Any, *, label: str) -> Path:
    if (not isinstance(relative, str) or not relative or len(relative) > 1024 or relative.strip() != relative
        or "\\" in relative or ":" in relative
        or any(ord(char) < 32 or ord(char) == 127 for char in relative)
        or any(part in {"", ".", "..", "node_modules", ".git", ".harbor-runtime"} for part in relative.split("/"))
        or PurePosixPath(relative).is_absolute()):
        _invalid(f"{label} must be a safe Candidate-relative source file")
    current = root
    for part in relative.split("/"):
        current = current / part
        if current.is_symlink():
            _invalid(f"{label} must not traverse symlinks")
    if not current.is_file() or not current.resolve().is_relative_to(root):
        _invalid(f"{label} must be an existing Candidate source file")
    return current


def _json(path: Path) -> dict[str, Any]:
    if path.stat().st_size > 4 * 1024 * 1024:
        _invalid(f"{path.name} exceeds the runtime metadata size limit")
    try:
        value = json.loads(path.read_text(encoding="utf-8"), parse_constant=lambda _: _invalid("runtime metadata must use standard JSON numbers"))
    except (OSError, UnicodeError, ValueError) as error:
        _invalid(f"{path.name} is not valid UTF-8 JSON: {type(error).__name__}")
    if not isinstance(value, dict):
        _invalid(f"{path.name} must be an object")
    return value


def _sha(path: Path) -> str:
    return "sha256:" + hashlib.sha256(path.read_bytes()).hexdigest()


def _locked_dependencies(package: dict[str, Any], lock: dict[str, Any]) -> None:
    packages = lock.get("packages")
    if lock.get("lockfileVersion") != 3 or not isinstance(packages, dict) or not isinstance(packages.get(""), dict):
        _invalid("execution requires package-lock.json v3 with its root package")
    locked_root = packages[""]
    for key in ("name", "version"):
        if not isinstance(package.get(key), str) or not package[key] or locked_root.get(key) != package[key]:
            _invalid(f"package-lock.json root {key} must match package.json")
    for kind in _DEPENDENCIES:
        declared = package.get(kind, {})
        if not isinstance(declared, dict) or locked_root.get(kind, {}) != declared:
            _invalid(f"package-lock.json root {kind} must match package.json")
        for name, version in declared.items():
            if not isinstance(version, str) or not _SEMVER.fullmatch(version):
                _invalid(f"root {kind} must use exact package versions, not tags or ranges")
            target = packages.get(f"node_modules/{name}")
            if not isinstance(target, dict) or target.get("version") != version:
                _invalid(f"root {kind} dependency must resolve to its exact locked version")
    for relative, entry in packages.items():
        if relative == "":
            continue
        if (not isinstance(relative, str) or not relative.startswith("node_modules/")
            or "\\" in relative or ":" in relative
            or any(ord(char) < 32 or ord(char) == 127 for char in relative)
            or any(part in {"", ".", "..", ".git", ".harbor-runtime"} for part in relative.split("/"))
            or not isinstance(entry, dict) or entry.get("link") or entry.get("inBundle")):
            _invalid("lockfile must contain registry packages, not local links or bundled entries")
        if not isinstance(entry.get("version"), str) or not _SEMVER.fullmatch(entry["version"]):
            _invalid("every locked package must have an exact version")
        resolved = entry.get("resolved")
        if (not isinstance(resolved, str) or not resolved.startswith("https://") or any(char in resolved for char in "\\?#")
            or any(ord(char) <= 32 or ord(char) == 127 for char in resolved)):
            _invalid("every locked package requires a credential-free HTTPS archive URL")
        try:
            url = urlsplit(resolved)
            # Accessing port also validates malformed or out-of-range values.
            url.port
        except ValueError:
            _invalid("locked package archive URL is invalid")
        if url.scheme != "https" or not url.hostname or url.username or url.password or url.query or url.fragment:
            _invalid("every locked package requires a credential-free HTTPS archive URL")
        integrity = entry.get("integrity")
        try:
            decoded = base64.b64decode(integrity[7:], validate=True) if isinstance(integrity, str) and integrity.startswith("sha512-") else b""
        except ValueError:
            decoded = b""
        if len(decoded) != 64 or base64.b64encode(decoded).decode() != integrity[7:]:
            _invalid("every locked package requires a SHA-512 archive integrity")


def load_candidate_runtime(candidate_dir: Path, *, required: bool = False) -> dict[str, Any]:
    root = candidate_dir.expanduser().resolve(strict=True)
    descriptor_path = root / DESCRIPTOR_NAME
    if not descriptor_path.exists() and not descriptor_path.is_symlink():
        if required:
            raise ValueError("CANDIDATE_RUNTIME_UNBOUND: this historical Candidate has no candidate-runtime.json; create a new Candidate with a locked local ACP entrypoint, then establish a fresh baseline. No demo/latest fallback is permitted.")
        return {"kind": "deepseek-harness", "policy": "unbound", "transport": "acp"}
    descriptor_path = _file(root, DESCRIPTOR_NAME, label="runtime descriptor")
    descriptor = _json(descriptor_path)
    if set(descriptor) != _FIELDS or isinstance(descriptor.get("schema_version"), bool) or descriptor.get("schema_version") != 1 or descriptor.get("transport") != "acp":
        _invalid("candidate-runtime.json requires the supported v1 ACP fields only")
    entrypoint = _file(root, descriptor.get("entrypoint"), label="entrypoint")
    if entrypoint.suffix not in {".js", ".mjs", ".cjs"}:
        _invalid("entrypoint must be a Candidate-owned Node.js source file")
    _file(root, descriptor.get("config_path"), label="config_path")
    entry_id = descriptor.get("agent_entry_id")
    if not isinstance(entry_id, str) or not _ENTRY_ID.fullmatch(entry_id):
        _invalid("agent_entry_id must be a bounded Cordis entry ID")
    node_version = descriptor.get("node_version")
    if not isinstance(node_version, str) or not _NODE_VERSION.fullmatch(node_version) or int(node_version.split(".")[0]) < 22:
        _invalid("node_version must be an exact Node.js version >= 22")
    package = _json(_file(root, "package.json", label="package manifest"))
    lockfile = _file(root, LOCKFILE_NAME, label="npm lockfile")
    _locked_dependencies(package, _json(lockfile))
    return {
        "kind": "deepseek-harness", "policy": "candidate-locked", "transport": "acp",
        "descriptor": DESCRIPTOR_NAME,
        **{key: descriptor[key] for key in ("entrypoint", "config_path", "agent_entry_id", "node_version")},
        "lockfile": LOCKFILE_NAME,
        "descriptor_digest": _sha(descriptor_path),
        "entrypoint_digest": _sha(entrypoint),
        "lockfile_digest": _sha(lockfile),
    }

File: apps/desktop-tauri/product/harbor-evolution/lib/candidate-runtime.js
import { createHash } from 'node:crypto'
import { lstat, readFile } from 'node:fs/promises'
import path from 'node:path'

export const CANDIDATE_RUNTIME_NAME = 'candidate-runtime.json'
const DESCRIPTOR_KEYS = ['schema_version', 'transport', 'entrypoint', 'config_path', 'agent_entry_id', 'node_version']
const RESERVED_PATHS = new Set(['node_modules', '.harbor-runtime', '.git'])
const EXACT_VERSION = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*)(?:\.(?:0|[1-9]\d*|\d*[A-Za-z-][0-9A-Za-z-]*))*)?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$/
const DEPENDENCY_FIELDS = ['dependencies', 'optionalDependencies', 'devDependencies', 'peerDependencies']

const object = value => value !== null && typeof value === 'object' && !Array.isArray(value)
const sha256 = bytes => `sha256:${createHash('sha256').update(bytes).digest('hex')}`

function safeRelative(value, label) {
  if (typeof value !== 'string' || !value || value.length > 1024 || value.trim() !== value || /[\\:\x00-\x1f\x7f]/.test(value) || path.posix.isAbsolute(value)) {
    throw new Error(`${label} must be a safe Candidate-relative path`)
  }
  const parts = value.split('/')
  if (parts.some(part => !part || part === '.' || part === '..' || RESERVED_PATHS.has(part))) {
    throw new Error(`${label} must not traverse or use reserved Candidate paths`)
  }
  return parts
}

async function sourceFile(root, relative, { optional = false } = {}) {
  const parts = safeRelative(relative, relative)
  let absolute = root
  for (let index = 0; index < parts.length; index++) {
    absolute = path.join(absolute, parts[index])
    let info
    try {
      info = await lstat(absolute)
    } catch (error) {
      if (optional && index === parts.length - 1 && error.code === 'ENOENT') return undefined
      if (error.code === 'ENOENT') throw new Error(`Candidate runtime requires a regular source file: ${relative}`)
      throw error
    }
    if (info.isSymbolicLink() || (index === parts.length - 1 ? !info.isFile() : !info.isDirectory())) {
      throw new Error(`Candidate runtime requires a regular source file without symlinks: ${relative}`)
    }
  }
  return readFile(absolute)
}

function parseObject(bytes, label) {
  if (bytes.length > 4 * 1024 * 1024) throw new Error(`${label} exceeds the runtime metadata size limit`)
  let value
  try {
    const source = bytes.toString('utf8')
    if (!Buffer.from(source, 'utf8').equals(bytes)) throw new Error('Invalid UTF-8')
    value = JSON.parse(source)
  } catch {
    throw new Error(`${label} is not valid JSON`)
  }
  if (!object(value)) throw new Error(`${label} must be an object`)
  return value
}

function dependencies(value, field, label) {
  const entries = Object.hasOwn(value, field) ? value[field] : {}
  if (!object(entries)) throw new Error(`${label} ${field} must be an object`)
  for (const [name, version] of Object.entries(entries)) {
    if (typeof version !== 'string' || !EXACT_VERSION.test(version)) {
      throw new Error(`${label} ${field}.${name} must use an exact semver, not a tag, range, alias, or local dependency`)
    }
  }
  return Object.entries(entries).sort(([left], [right]) => left.localeCompare(right))
}

function validateLockfile(packageJson, lockfile) {
  if (lockfile.lockfileVersion !== 3 || !object(lockfile.packages) || !object(lockfile.packages[''])) {
    throw new Error('Candidate runtime requires package-lock.json v3 with packages[\'\'] root metadata')
  }
  const root = lockfile.packages['']
  for (const field of ['name', 'version']) {
    if (typeof packageJson[field] !== 'string' || !packageJson[field] || root[field] !== packageJson[field]) {
      throw new Error(`package-lock.json root ${field} must match package.json ${field}`)
    }
  }
  for (const field of DEPENDENCY_FIELDS) {
    const expected = dependencies(packageJson, field, 'package.json')
    const actual = dependencies(root, field, 'package-lock.json root')
    if (JSON.stringify(expected) !== JSON.stringify(actual)) {
      throw new Error(`package-lock.json root ${field} must match package.json ${field}`)
    }
    for (const [name, version] of expected) {
      const target = lockfile.packages[`node_modules/${name}`]
      if (!object(target) || target.version !== version) {
        throw new Error(`package-lock.json root ${field} dependency must resolve to its exact locked version`)
      }
    }
  }
  for (const [location, entry] of Object.entries(lockfile.packages)) {
    if (!location) continue
    if (!location.startsWith('node_modules/') || /[\\:\x00-\x1f\x7f]/.test(location) || location.split('/').some(part => !part || part === '.' || part === '..' || part === '.git' || part === '.harbor-runtime')) {
      throw new Error(`package-lock.json contains an unsupported package location: ${location}`)
    }
    if (!object(entry) || entry.link || entry.inBundle || typeof entry.version !== 'string' || !EXACT_VERSION.test(entry.version)) {
      throw new Error(`package-lock.json ${location} requires an exact registry version without links or bundled dependencies`)
    }
    let resolved
    try {
      resolved = new URL(entry.resolved)
    } catch {
      throw new Error(`package-lock.json ${location} requires an HTTPS resolved package URL`)
    }
    if (typeof entry.resolved !== 'string' || !entry.resolved.startsWith('https://') || /[\x00-\x20\x7f\\]/.test(entry.resolved) || resolved.protocol !== 'https:' || !resolved.hostname || resolved.username || resolved.password || resolved.search || resolved.hash || entry.resolved.includes('?') || entry.resolved.includes('#')) {
      throw new Error(`package-lock.json ${location} requires an HTTPS resolved package URL without credentials, query, or fragment`)
    }
    const integrity = entry.integrity
    if (typeof integrity !== 'string' || !/^sha512-[A-Za-z0-9+/]{86}==$/.test(integrity) || Buffer.from(integrity.slice(7), 'base64').toString('base64') !== integrity.slice(7)) {
      throw new Error(`package-lock.json ${location} requires a valid sha512 integrity digest`)
    }
  }
}

/** Resolve only Candidate-owned, source-digested runtime metadata; never select a registry default. */
export async function loadCandidateRuntime(candidateDir, { required = false } = {}) {
  const root = path.resolve(candidateDir)
  const bytes = await sourceFile(root, CANDIDATE_RUNTIME_NAME, { optional: true })
  if (bytes === undefined) {
    if (required) throw new Error(`Candidate runtime is unbound; migrate this Candidate by adding ${CANDIDATE_RUNTIME_NAME} and a locked ACP entrypoint, then create a new snapshot`)
    return { kind: 'deepseek-harness', policy: 'unbound', transport: 'acp' }
  }
  const descriptor = parseObject(bytes, CANDIDATE_RUNTIME_NAME)
  if (Object.keys(descriptor).some(key => !DESCRIPTOR_KEYS.includes(key)) || DESCRIPTOR_KEYS.some(key => !Object.hasOwn(descriptor, key))) {
    throw new Error(`${CANDIDATE_RUNTIME_NAME} requires exactly: ${DESCRIPTOR_KEYS.join(', ')}`)
  }
  if (descriptor.schema_version !== 1 || descriptor.transport !== 'acp') {
    throw new Error(`${CANDIDATE_RUNTIME_NAME} requires schema_version=1 and transport=acp`)
  }
  safeRelative(descriptor.entrypoint, 'entrypoint')
  if (!/\.(?:js|mjs|cjs)$/.test(descriptor.entrypoint)) throw new Error('Candidate runtime entrypoint must be a .js, .mjs, or .cjs source file')
  safeRelative(descriptor.config_path, 'config_path')
  if (typeof descriptor.agent_entry_id !== 'string' || !/^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/.test(descriptor.agent_entry_id)) {
    throw new Error('Candidate runtime agent_entry_id must be a non-empty safe identifier')
  }
  if (typeof descriptor.node_version !== 'string' || !/^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/.test(descriptor.node_version) || Number(descriptor.node_version.split('.')[0]) < 22) {
    throw new Error('Candidate runtime node_version must be an exact x.y.z release with major >=22')
  }
  const entrypointBytes = await sourceFile(root, descriptor.entrypoint)
  await sourceFile(root, descriptor.config_path)
  const packageBytes = await sourceFile(root, 'package.json')
  const lockfileBytes = await sourceFile(root, 'package-lock.json')
  validateLockfile(parseObject(packageBytes, 'package.json'), parseObject(lockfileBytes, 'package-lock.json'))
  return {
    kind: 'deepseek-harness',
    policy: 'candidate-locked',
    transport: 'acp',
    descriptor: CANDIDATE_RUNTIME_NAME,
    entrypoint: descriptor.entrypoint,
    config_path: descriptor.config_path,
    agent_entry_id: descriptor.agent_entry_id,
    node_version: descriptor.node_version,
    lockfile: 'package-lock.json',
    descriptor_digest: sha256(bytes),
    entrypoint_digest: sha256(entrypointBytes),
    lockfile_digest: sha256(lockfileBytes),
  }
}
